### PR TITLE
Solver find solutions method

### PR DESCRIPTION
**Pull request by IronicJuice**
Added Solver.FindSolutions(), which returns all, or a certain number of solutions to a given board. Can also randomize.

Also added Board.Validate with tests

There are several known improvements that can be made, but these will be part of a larger performance test of the solver.

### CODE REVIEW

File: SudoScript.Core.Test/BoardTests.cs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        board[9, 9].EliminateCandidate(3);
+        board[9, 9].EliminateCandidate(4);
+        board[9, 9].EliminateCandidate(5);
+        board[9, 9].EliminateCandidate(6);
+        board[9, 9].EliminateCandidate(7);
+        board[9, 9].EliminateCandidate(8);
+        board[9, 9].EliminateCandidate(9);
+
+        Assert.That(board.Validate(), Is.False);
+    }
```
Comment (frederikja163) [inline SudoScript.Core.Test/BoardTests.cs]: This is multiple different unit tests in one. This is the reason we usually use AAA for unit tests (arrange act assert) which this unit test definitely does not follow.

File: SudoScript.Core/Solver.cs [hunk 2]
```diff
@@ -71,14 +72,96 @@ private static bool SolveRec(Board board, [NotNullWhen(true)] out Board? solvedB
...
     }
 
+    /// <summary>
+    /// Finds a number of solutions to a given board.
+    /// </summary>
+    /// <param name="board"></param>
+    /// <param name="random">Whether the solved boards should be randomized.</param>
+    /// <param name="solutionCount">Maximum number of solutions that should be in the returned list. Zero means unlimited.</param>
+    /// <returns></returns>
+    public static List<Board>? FindSolutions(Board board, int limit = 0, bool random = false)
```
Comment (frederikja163) [inline SudoScript.Core/Solver.cs]: I would reccommend against using magic numbers as you do with 0 here if at all possible. If you instead set limit to int.MaxValue you will also ensure to not generate more solutions than can be stored in a list.

However, i dont get why we are even storing these in a list at all. Seems like it would both be simpler and safer to store them directly in an IEnumerable. Then the user can just say .Take themselves.

Comment (frederikja163) [reply SudoScript.Core/Solver.cs]: Upon looking a bit further, since you are creating these in a hashset returning an IEnumerable might not be the best idea anyways. If there is a way to do it that would be good. But it would make duplicate detection a bigger problem. If you can figure out how to do it that would be really nice. Otherwise you can go ahead with a list. (You could also consider returning the hashset directly which might honestly be my prefered option.

Still change the 0 to be int.MaxValue though.

File: SudoScript.Core/Solver.cs [hunk 3]
```diff
@@ -71,14 +72,96 @@ private static bool SolveRec(Board board, [NotNullWhen(true)] out Board? solvedB
...
+    {
+        HashSet<Board> result = SolveRecAll(board, 0, limit, random);
+        // If a limit is specified and the result count exceeds it, return only the correct number of solutions.
+        if (limit != 0 && result.Count > limit)
+        {
+            return result.Take(limit).ToList();
```
Comment (frederikja163) [inline SudoScript.Core/Solver.cs:88]: Is there a reason to take a limit here? Shouldn't SolveRecAll already give you the limit?

File: SudoScript.Core/Solver.cs [hunk 4]
```diff
@@ -71,14 +72,96 @@ private static bool SolveRec(Board board, [NotNullWhen(true)] out Board? solvedB
...
         throw new NotImplementedException();
     }
 
-    public static bool IsSatisfactory(Board board)
+    /// <summary>
+    /// Checks if the board can be solved by just using the EliminateCandidates methods from units.
+    /// </summary>
+    /// <param name="board"></param>
+    /// <returns>True if the board can be solved without trial and error guessing.</returns>
+    public static bool IsSatisfactory(Board board) // Certain methods for eliminating candidates using inference are not currently implemented. Implementing them would make this function more acurate.
```
Comment (frederikja163) [inline SudoScript.Core/Solver.cs:158]: Wouldn't it be better to check this by checking all solutions and seeing if there is only a single solution? If there is only a single solution that solution could theoretically always be infered (by simulating infinite backtracking which could be done even on paper).

**Review by frederikja163 — CHANGES_REQUESTED** [review]
Overall this is looking good. A few minor changes requested.

Also (and not sure if this is applicable) didnt we already implement a solver method? Is there any reason you cant extend the capabilities of the current method instead of adding a new method?

**Comment by frederikja163** [comment]
Would be nice if you could mention the known improvements here for further documentation. Also when opening pull requests without an issue attached listing the reasons for the changes could help when reviewing.

File: SudoScript.Core/Solver.cs (hunk 3 above)
Comment (IronicJuice) [reply SudoScript.Core/Solver.cs:88]: This is to avoid edge cases where the list contains one too many solutions. I'm not sure if these cases exist, but I'm not sure how to properly test for it

File: SudoScript.Core/Solver.cs (hunk 4 above)
Comment (IronicJuice) [reply SudoScript.Core/Solver.cs:158]: I believe you may be confusing satisfactory with proper

**Comment by IronicJuice** [comment]
The purpose of the FindSolutions method, is to allow other wanted features late. IsProper() can be infered by asking if FingSolutions() returns more than one solution. These found solutions can also be used to create boards.

I am working on a separate branch to test improvements such as:

- Smarter candidate selection using heuristics
- Parallel solving (multithreading)
- More aggressive candidate pruning
and any other I might stumble across as long as I can justify them with increased performance or other metrics.

File: SudoScript.Core.Test/BoardTests.cs (hunk 1 above)
Comment (IronicJuice) [reply SudoScript.Core.Test/BoardTests.cs]: Addressed with [5a73406]

File: SudoScript.Core/Solver.cs (hunk 2 above)
Comment (IronicJuice) [reply SudoScript.Core/Solver.cs]: Adressed with [73a7bd8].
I will look into returning in a different way as I test these following days. This pull request stands for now.

File: SudoScript.Core/Solver.cs (hunk 4 above)
Comment (frederikja163) [reply SudoScript.Core/Solver.cs:158]: Those two terms actually mean the same as i explained in my comment.

If there is only one solution you can always reach said solution without guess work. If you cannot reach the one solution without guess work that means there are more solutions.

File: SudoScript.Core/Solver.cs (hunk 3 above)
Comment (frederikja163) [reply SudoScript.Core/Solver.cs:88]: Then i think you should try to gain an understanding of that. A test like this is really error prone and probably not even needed.
